providers/googledns/googledns.go

```diff
@@ -0,0 +1,197 @@
+//   4. On Google Compute Engine and Google App Engine Managed VMs, it fetches
+//      credentials from the metadata server.
+//      (In this final case any provided scopes are ignored.)
+
+func (r *GoogleDNSProvider) Init(rootDomainName string) error {
+	r.limiter = ratelimit.NewBucketWithRate(5.0, 1)
+
+	if envVal := os.Getenv("GOOGLEDNS_MAX_RETRIES"); envVal != "" {
+		i, err := strconv.Atoi(envVal)
+		if err == nil {
+			googlednsMaxRetries = i
+		} else {
+			logrus.Warnf("Invalid value for GOOGLEDNS_MAX_RETRIES. Using default.")
+		}
+	}
+
+	ctx := cnet.Context(cnet.Background())
+
+	c, _ := google.DefaultClient(ctx, dns.CloudPlatformScope)
+
+	r.client, _ = dns.New(c)
+
+	if err := r.setHostedZone(rootDomainName); err != nil {
+		return fmt.Errorf("Failed to configure hosted zone: %v", err)
+	}
+
+	logrus.Infof("Configured %s with hosted zone %s",
+		r.GetName(), rootDomainName)
+
+	return nil
+}
+
+func (r *GoogleDNSProvider) setHostedZone(rootDomainName string) error {
+	r.limiter.Wait(1)
+
+	zones, err := r.client.ManagedZones.List(projectId).Do()
+
+	if len(zones.ManagedZones) == 0 {
+		return fmt.Errorf("Hosted zone for '%s' not found, got empty list", rootDomainName)
+	}
+
+	for _, zone := range zones.ManagedZones {
+		fmt.Printf("Found Zone: %v (%v)\n", zone.Name, zone.DnsName)
+		if zone.DnsName == rootDomainName {
+			fmt.Printf("Found desired zone %v !", zone.Name)
+			r.hostedZoneId = string(zone.Name)
+			return nil
+		}
+	}
+
+	return err
+
+}
+
+func (*GoogleDNSProvider) GetName() string {
+	return "Google Cloud DNS"
+}
+
+func (r *GoogleDNSProvider) HealthCheck() error {
+	return nil
+}
+
+func (r *GoogleDNSProvider) AddRecord(record utils.DnsRecord) error {
+	return r.changeRecord(record, "ADD")
+}
+
+func (r *GoogleDNSProvider) UpdateRecord(record utils.DnsRecord) error {
+	return r.changeRecord(record, "UPDATE")
+}
+
+func (r *GoogleDNSProvider) RemoveRecord(record utils.DnsRecord) error {
+	return r.changeRecord(record, "DELETE")
+}
+
+func (r *GoogleDNSProvider) changeRecord(record utils.DnsRecord, action string) error {
+	r.limiter.Wait(1)
+
+	var records []*dns.ResourceRecordSet
+
+	rec := dns.ResourceRecordSet{
+		Kind:            "dns#resourceRecordSet",
+		Name:            record.Fqdn,
+		Rrdatas:         record.Records,
+		Ttl:             int64(record.TTL),
+		Type:            record.Type,
+		ForceSendFields: nil,
+		NullFields:      nil,
+	}
+
+	records = append(records, &rec)
+
+	change := dns.Change{
+		Additions: nil,
+		Deletions: nil,
+		Kind:      "dns#change",
+	}
+
+	if action == "ADD" {
+		change.Additions = records
+	} else if action == "DELETE" || action == "UPDATE" {
+
+		var recs []*dns.ResourceRecordSet
+
+		rrSets, _ := r.client.ResourceRecordSets.List(projectId, r.hostedZoneId).Do()
+
+		for _, rrSet := range rrSets.Rrsets {
+
+			if rrSet.Name == record.Fqdn {
+				recs = append(recs, rrSet)
+			}
+		}
+
+		if action == "UPDATE" {
+			change.Additions = records
+		}
+		change.Deletions = recs
```

Comment:
for action UPDATE, do we also populate change.Deletions?

Reply:
Yes. There is no update method on the API, so in the same batch we have to remove the existing ones and add the updated entries.